foreman/data_refinery_foreman/foreman/management/commands/create_compendia.py

```diff
@@ -65,33 +35,143 @@ def add_arguments(self, parser):
+    result = []
+
+    # process the organisms with the most platforms first
+    target_organisms_sorted = target_organisms\
+        .annotate(num_platforms=Count('sample__platform_accession_code', distinct=True))\
+        .order_by('-num_platforms')
+
+    for organism in target_organisms_sorted:
+        added_with_another_organism = any(
+            any(item == organism for item in group) for group in result)
+        if added_with_another_organism:
+            continue
+
+        organism_group = [organism]
+
+        platform_list = list(get_organism_microarray_platforms(organism))
+
+        genus_prefix = organism.get_genus() + '_'
+        organisms_sharing_genus = Organism.objects\
+            .exclude(id=organism.id)\
+            .filter(name__startswith=genus_prefix)
+        for genus_organism in organisms_sharing_genus:
+            # Group together the orgenisms that share at least one processed
+            # sample in a microarray platform
+            platforms = get_organism_microarray_platforms(genus_organism)\
+                .filter(platform_accession_code__in=platform_list)
+            if platforms.exists():
+                organism_group.append(genus_organism)
+
+        result.append(organism_group)
+
+    return result
+
+
+def get_organism_microarray_platforms(organism):
+    """ Returns the accession codes of the Affymetrix microarray platforms associated with an
+    organism. Ordered by the number of samples for each platform in descending order """
+    return organism.sample_set.filter(has_raw=True, technology="MICROARRAY", is_processed=True)\
+                   .values('platform_accession_code')\
+                   .annotate(count=Count('id'))\
+                   .order_by('-count')\
+                   .values_list('platform_accession_code', flat=True)
+
+
+def create_job_for_organism(organisms: List[Organism], quant_sf_only=False, svd_algorithm='ARPACK'):
+    """Returns a compendia job for the provided organism.
+
+    Fetch all of the experiments and compile large but normally formated Dataset.
+    """
+    job = ProcessorJob()
+    job.pipeline_applied = ProcessorPipeline.CREATE_COMPENDIA.value
+    job.save()
+
+    dataset = Dataset()
+    dataset.data = get_dataset(organisms)
+    dataset.scale_by = 'NONE'
+    # The quantpendias should be aggregated by species
+    dataset.aggregate_by = 'EXPERIMENT' if quant_sf_only else 'SPECIES'
+    dataset.quantile_normalize = False
+    dataset.quant_sf_only = quant_sf_only
+    dataset.svd_algorithm = svd_algorithm
+    dataset.save()
+
+    pjda = ProcessorJobDatasetAssociation()
+    pjda.processor_job = job
+    pjda.dataset = dataset
+    pjda.save()
+
+    return job
+
+
+def get_dataset(organisms: List[Organism]):
+    """ Builds a dataset with the samples associated with the given organisms """
+    dataset = {}
+
+    filter_query = Q()
+    for organism in organisms:
+        filter_query = filter_query | Q(organisms=organism)
```

Comment:
Is this the right way to query a many-to-many field? If it is cool it's just that `organisms=organism` looks like we're comparing a list to a single object so it should never be true, but I could see that just being how this is done in Django.

Reply:
Yes, it was strange to me as well. I have seen it in several places and they use it in the [docs](https://docs.djangoproject.com/en/2.2/topics/db/examples/many_to_many/) as well.